src/adapter/src/explain_new/optimizer_trace.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Tracing utilities for explainable plans.
+
+use std::fmt::Debug;
+
+use mz_compute_client::{command::DataflowDescription, plan::Plan};
+use mz_expr::{MirRelationExpr, MirScalarExpr, OptimizedMirRelationExpr, RowSetFinishing};
+use mz_ore::tracing::{PlanTrace, TraceEntry};
+use mz_repr::{
+    explain_new::{text_string, DisplayText, Explain, ExplainConfig, ExplainError, ExplainFormat},
+    Timestamp,
+};
+use mz_sql::plan::{HirRelationExpr, HirScalarExpr};
+use tracing::dispatcher::{self, with_default};
+use tracing_subscriber::prelude::*;
+
+use crate::{catalog::ConnCatalog, coord::peek::FastPathPlan};
+
+use super::{Displayable, ExplainContext, Explainable, UsedIndexes};
+
+type LirRelationExpr = Plan<Timestamp>;
+
+/// Provides functionality for tracing plans generated by the execution of an
+/// optimization pipeline.
+pub(crate) struct OptimizerTrace(dispatcher::Dispatch);
+
+impl OptimizerTrace {
+    pub fn new() -> OptimizerTrace {
+        let subscriber = tracing_subscriber::registry()
```

Comment:
Because we have a lot of different plan types `T` the only way to collect the trace at the moment is to enumerate all of them as a chain of layered `PlanTrace<T>` listeners, one for each plan.

Once `Valuable` is stabilized I can probably do this in a more efficient way.

Reply:
this seems reasonable for now as a code organization technique, instead of making `PlanTrace` have to match on all them!